File: client/app/src/main/java/me/urakalee/next2/model/Note.java
package me.urakalee.next2.model;

import android.net.Uri;
import android.support.annotation.NonNull;

import org.joda.time.LocalDate;

import java.io.File;

import me.shouheng.notepal.model.Model;
import me.shouheng.notepal.model.Notebook;
import me.shouheng.notepal.provider.annotation.Column;
import me.shouheng.notepal.provider.annotation.Table;
import me.shouheng.notepal.provider.schema.NoteSchema;
import me.urakalee.next2.config.TimeConfig;

/**
 * Created by wangshouheng on 2017/5/12.
 */
@Table(name = NoteSchema.TABLE_NAME)
public class Note extends Model {

    public static final String DEFAULT_SUFFIX = ".md";

    private static final String DAY_SEPARATOR = "_";

    @Column(name = NoteSchema.TREE_PATH)
    private String treePath;

    @Column(name = NoteSchema.TITLE)
    private String title;

    @Column(name = NoteSchema.CONTENT_CODE)
    private long attachmentCode;

    @Column(name = NoteSchema.TAGS)
    private String tags;

    @Column(name = NoteSchema.PREVIEW_IMAGE)
    private Uri previewImage;

    @Column(name = NoteSchema.PREVIEW_CONTENT)
    private String previewContent;

    // region Android端字段，不计入数据库

    private Notebook notebook;
    private String content;
    private String tagsName;

    private String timePath;
    private int dayPrefix;
    private String originTitle;
    private File originFile;

    private String getFileNameWithDayPrefix(String name) {
        if (dayPrefix > 0) {
            return String.format("%02d%s%s", dayPrefix, DAY_SEPARATOR, name);
        } else {
            return name;
        }
    }

    public String getFileName() {
        String result;
        if (title.endsWith(DEFAULT_SUFFIX)) {
            result = title;
        } else {
            result = title + DEFAULT_SUFFIX;
        }
        return getFileNameWithDayPrefix(result);
    }

    public void setTitleByFileName(@NonNull String name) {
        String result;
        if (name.endsWith(DEFAULT_SUFFIX)) {
            result = name.substring(0, name.length() - DEFAULT_SUFFIX.length());
        } else {
            result = name;
        }
        String[] parts = result.split(DAY_SEPARATOR, 2);
        if (parts.length == 2) {
            try {
                setDayPrefix(Integer.parseInt(parts[0]));
                title = parts[1];
                return;
            } catch (NumberFormatException e) {
                // ignore
            }
        }
        setDayPrefix(0);
        title = result;
    }

    public Notebook getNotebook() {
        return notebook;
    }

    public void setNotebook(Notebook notebook) {
        this.notebook = notebook;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public String getTagsName() {
        return tagsName;
    }

    public void setTagsName(String tagsName) {
        this.tagsName = tagsName;
    }

    public String getTimePath() {
        return timePath;
    }

    public void setTimePath(String timePath) {
        this.timePath = timePath;
    }

    public int getDayPrefix() {
        return dayPrefix;
    }

    public void setDayPrefix(int day) {
        this.dayPrefix = day;
    }

    public void generateTimePath() {
        LocalDate today = LocalDate.now();
        setTimePath(today.toString(TimeConfig.MONTH_FORMAT));
        setDayPrefix(today.dayOfMonth().get());
    }

    public String getCreateTime() {
        if (dayPrefix > 0) {
            // 验证 dayPrefix 的有效性
            LocalDate firstDay = LocalDate.parse(String.format("%s-01", timePath));
            int month = firstDay.getMonthOfYear();
            LocalDate createDay = firstDay.plusDays(dayPrefix - 1);
            if (createDay.getMonthOfYear() == month) {
                return createDay.toString("yyyy-MM-dd");
            } else {
                return timePath;
            }
        } else {
            return timePath;
        }
    }

    public String getOriginTitle() {
        return originTitle;
    }

    public String getOriginFileName() {
        String result;
        if (originTitle.endsWith(DEFAULT_SUFFIX)) {
            result = originTitle;
        } else {
            result = originTitle + DEFAULT_SUFFIX;
        }
        return getFileNameWithDayPrefix(result);
    }

    public void setOriginTitle(String originTitle) {
        this.originTitle = originTitle;
    }

    public boolean isNewNote() {
        return originTitle == null;
    }

    public void finishNew() {
        originTitle = title;
    }

    public boolean needRename() {
        return originTitle != null && !title.equals(originTitle);
    }

    public File getOriginFile() {
        return originFile;
    }

    public void setOriginFile(File originFile) {
        this.originFile = originFile;
    }

    public void finishRename() {
        originTitle = title;
        originFile = null;
    }

    // endregion

    public Note() {
    }

    public String getTreePath() {
        return treePath;
    }

    public void setTreePath(String treePath) {
        this.treePath = treePath;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public long getAttachmentCode() {
        return attachmentCode;
    }

    public void setAttachmentCode(long attachmentCode) {
        this.attachmentCode = attachmentCode;
    }

    public String getTags() {
        return tags;
    }

    public void setTags(String tags) {
        this.tags = tags;
    }

    public Uri getPreviewImage() {
        return previewImage;
    }

    public void setPreviewImage(Uri previewImage) {
        this.previewImage = previewImage;
    }

    public String getPreviewContent() {
        return previewContent;
    }

    public void setPreviewContent(String previewContent) {
        this.previewContent = previewContent;
    }

    @Override
    public String toString() {
        return "Note{" +
                "treePath='" + treePath + '\'' +
                ", title='" + title + '\'' +
                ", attachmentCode=" + attachmentCode +
                ", tags='" + tags + '\'' +
                ", previewImage=" + previewImage +
                ", previewContent='" + previewContent + '\'' +
                ", content='" + content + '\'' +
                ", tagsName='" + tagsName + '\'' +
                "} " + super.toString();
    }
}

File: client/app/src/main/java/me/urakalee/next2/model/Note.kt
package me.urakalee.next2.model

import android.net.Uri
import me.shouheng.notepal.model.Model
import me.shouheng.notepal.model.Notebook
import me.shouheng.notepal.provider.annotation.Column
import me.shouheng.notepal.provider.annotation.Table
import me.shouheng.notepal.provider.schema.NoteSchema
import me.urakalee.next2.config.TimeConfig
import me.urakalee.next2.storage.getFile
import org.joda.time.LocalDate
import java.io.File

/**
 * @author Uraka.Lee
 */
@Table(name = NoteSchema.TABLE_NAME)
class Note : Model() {

    @Column(name = NoteSchema.TREE_PATH)
    var treePath: String? = null

    @Column(name = NoteSchema.TITLE)
    var title: String? = null

    @Column(name = NoteSchema.CONTENT_CODE)
    var attachmentCode: Long = 0

    @Column(name = NoteSchema.TAGS)
    var tags: String? = null

    @Column(name = NoteSchema.PREVIEW_IMAGE)
    var previewImage: Uri? = null

    @Column(name = NoteSchema.PREVIEW_CONTENT)
    var previewContent: String? = null

    // region Android端字段，不计入数据库

    var notebook: Notebook? = null
    var content: String? = null
    var tagsName: String? = null

    var timePath: String? = null
    var dayPrefix: Int = 0
    var originTitle: String? = null
    var originFile: File? = null

    //endregion

    val fileName: String?
        get() {
            val titleNonNull = title ?: return null
            return getFileNameWithDayPrefix(
                    if (titleNonNull.endsWith(DEFAULT_SUFFIX)) {
                        titleNonNull
                    } else {
                        titleNonNull + DEFAULT_SUFFIX
                    }
            )
        }

    val file: File?
        get() {
            val notebookNonNull = notebook ?: return null
            val timePathNonNull = timePath ?: return null
            val fileNameNonNull = fileName ?: return null
            return getFile(notebookNonNull.title, timePathNonNull, fileNameNonNull)
        }

    val createTimeStr: String
        get() {
            // 列表展示时使用, timePath 一定不空
            val timePathNonNull = timePath!!
            // 验证 dayPrefix 的有效性
            return if (dayPrefix > 0) {
                val firstDay = LocalDate.parse(String.format("%s-01", timePath))
                val month = firstDay.monthOfYear
                val createDay = firstDay.plusDays(dayPrefix - 1)
                return if (createDay.monthOfYear == month) {
                    createDay.toString("yyyy-MM-dd")
                } else {
                    timePathNonNull
                }
            } else {
                timePathNonNull
            }
        }

    //region load

    fun setTitleByFileName(name: String) {
        val result = if (name.endsWith(DEFAULT_SUFFIX)) {
            name.substring(0, name.length - DEFAULT_SUFFIX.length)
        } else {
            name
        }
        val parts = result.split(DAY_SEPARATOR, limit = 2)
        dayPrefix = if (parts.size == 2) {
            parts[0].toIntOrNull() ?: 0
        } else {
            0
        }
        title = if (dayPrefix > 0) parts[1] else result
    }

    //endregion
    //region save

    val isNewNote: Boolean
        get() = originTitle == null

    fun finishNew() {
        originTitle = title
    }

    fun needRename(): Boolean {
        return originTitle != null && title != originTitle
    }

    val originFileName: String
        get() {
            // 改名时使用, originTitle 一定不空
            val originTitleNonNull = originTitle!!
            return getFileNameWithDayPrefix(
                    if (originTitleNonNull.endsWith(DEFAULT_SUFFIX)) {
                        originTitleNonNull
                    } else {
                        originTitleNonNull + DEFAULT_SUFFIX
                    }
            )
        }

    fun finishRename() {
        originTitle = title
        originFile = null
    }

    fun generateTimePath() {
        val today = LocalDate.now()
        timePath = today.toString(TimeConfig.MONTH_FORMAT)
        dayPrefix = today.dayOfMonth().get()
    }

    //endregion
    //region private

    private fun getFileNameWithDayPrefix(name: String): String {
        return if (dayPrefix > 0) {
            String.format("%02d%s%s", dayPrefix, DAY_SEPARATOR, name)
        } else {
            name
        }
    }

    //endregion

    override fun toString(): String {
        return "Note{" +
                "treePath='" + treePath + '\''.toString() +
                ", title='" + title + '\''.toString() +
                ", attachmentCode=" + attachmentCode +
                ", tags='" + tags + '\''.toString() +
                ", previewImage=" + previewImage +
                ", previewContent='" + previewContent + '\''.toString() +
                ", content='" + content + '\''.toString() +
                ", tagsName='" + tagsName + '\''.toString() +
                "} " + super.toString()
    }

    companion object {

        const val DEFAULT_SUFFIX = ".md"

        private const val DAY_SEPARATOR = "_"
    }
}
